src/operator/nn/im2col.cu

```diff
@@ -0,0 +1,652 @@
+    // computations used to compute the spatial indices.
+    int channel_in = index;
+    int channel_out = 1;
+    for (i = num_axes - 1; i >= 0; --i) {
+      d_temp[i] = channel_in % shared_col_shape[i + 1];
+      channel_in /= shared_col_shape[i + 1];
+      channel_out *= shared_kernel_shape[i];
+    }
+    channel_out *= channel_in;
+    int data_col_inc = 1;
+    for (i = 0; i < num_axes; ++i) {
+      channel_out *= shared_col_shape[i + 1];
+      channel_out += d_temp[i];
+      d_temp[i] = d_temp[i] * shared_stride[i] - shared_pad[i];
+      channel_in *= shared_im_shape[i + 1];
+      channel_in += d_temp[i];
+      data_col_inc *= shared_col_shape[i + 1];
+      d_iter[i] = 0;
+    }
+    DType* data_col_ptr = data_col + channel_out;
+    const DType* data_im_ptr = data_im + channel_in;
+    bool incremented;
+    do {
+      bool in_range = true;
+      for (i = 0; i < num_axes; ++i) {
+        const int d_iter_im = d_iter[i] * shared_dilation[i] + d_temp[i];
+        in_range &= d_iter_im >= 0 && d_iter_im < shared_im_shape[i + 1];
+        if (!in_range) { break; }
+      }
+      if (in_range) {
+        int data_im_offset = d_iter[0] * shared_dilation[0];
+        for (i = 1; i < num_axes; ++i) {
+          data_im_offset *= shared_im_shape[i + 1];
+          data_im_offset += d_iter[i] * shared_dilation[i];
+        }
+        *data_col_ptr = data_im_ptr[data_im_offset];
+      } else {
+        *data_col_ptr = 0;
+      }
+      data_col_ptr += data_col_inc;
+      incremented = false;
+      for (i = num_axes - 1; i >= 0; --i) {
+        const int d_max = shared_kernel_shape[i];
+        if (d_iter[i] == d_max - 1) {
+          d_iter[i] = 0;
+        } else {  // d_iter[i] < d_max - 1
+          ++d_iter[i];
+          incremented = true;
+          break;
+        }
+      }  // for (int i = num_axes - 1; i >= 0; --i)
+    } while (incremented);  // do
+  }  // CUDA_KERNEL_LOOP(index, n)
+}
+
+// new implementation
+template <typename DType>
+void im2col_nd_gpu(mshadow::Stream<gpu>* s,
+    const DType* data_im, const int num_spatial_axes, const int num_kernels,
+    const TShape& im_shape, const TShape& col_shape,
+    const TShape& kernel_shape, const TShape& pad, const TShape& stride,
+    const TShape& dilation, DType* data_col) {
+  // num_axes should be smaller than block size
+  CHECK_LT(num_spatial_axes, mshadow::cuda::kBaseThreadNum);
+  switch (num_spatial_axes) {
+  case 1:
+    im2col_nd_gpu_kernel<DType, 1>  // NOLINT_NEXT_LINE(whitespace/operators)
+        <<<cuda_get_num_blocks(num_kernels), mshadow::cuda::kBaseThreadNum,
+           0, mshadow::Stream<gpu>::GetStream(s)>>>(
+        num_kernels, data_im, im_shape.get<3>(), col_shape.get<2>(),
+        kernel_shape.get<1>(), pad.get<1>(), stride.get<1>(), dilation.get<1>(), data_col);
+    break;
+  case 2:
+    im2col_nd_gpu_kernel<DType, 2>  // NOLINT_NEXT_LINE(whitespace/operators)
```

Comment:
You can merge im2col_nd_gpu and im2col_gpu by calling im2col_gpu_kernel for case 2 instead